iosApp/Modules/Helpers/Files/DownloadManager.swift

```diff
@@ -0,0 +1,78 @@
+import Foundation
+import Combine
+
+class DownloadManager: NSObject, ObservableObject {
+	@Published private(set) var downloads: [URL: URLSessionDownloadTask] = [:]
+	private var progressHandlers: [URL: (Double) -> Void] = [:]
+	
+	private var urlSession: URLSession!
+	private var downloadCompletionHandlers: [URL: (Result<URL, Error>) -> Void] = [:]
+	
+	override init() {
+		super.init()
+		urlSession = URLSession(configuration: .default, delegate: self, delegateQueue: OperationQueue.main)
+	}
+	
+	func download(url: URL, completion: @escaping (Result<URL, Error>) -> Void, progressHandler: @escaping (Double) -> Void) {
+		guard downloads[url] == nil else { return }
+		
+		let task = urlSession.downloadTask(with: url)
+		downloads[url] = task
+		downloadCompletionHandlers[url] = completion
+		progressHandlers[url] = progressHandler
+		task.resume()
+	}
+	
+	func download(url: URL, progressHandler: @escaping (Double) -> Void) async throws -> URL {
+		try await withCheckedThrowingContinuation { [weak self] continuation in
+			guard let self else { fatalError() }
+			download(url: url) { result in
+				continuation.resume(with: result)
+			} progressHandler: { progress in
+//				print("\n\n\nCANCELLED\n\n\n")
```

Comment:
I have not done the implementation for file download just yet, what would be the case where you cancel a download and start a new download? Are you talking about the user cancelling the download purposely?

Reply:
That's another case yeah, but i was thinking if they started another download before the current one is finished, it would cancel the current, so there was no queue.  But now that I think of it, that seems silly.  So I'll just remove this.